### PR TITLE
watcher: add supervisor to monitor watchers

### PR DESCRIPTION
**Pull request by panoel**

### CODE REVIEW

File: watcher/src/workers/supervisor.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  lastHB: number;
+}
+
+const workers: { [key: string]: WorkerInfo } = {};
+const logger = getLogger('supervisor');
+const network: Environment = getEnvironment();
+
+function spawnWorker(data: WorkerData) {
+  const workerName = `${data.chain}Worker`;
+  console.log('workerName:', workerName);
```
Comment (bingyuyap) [inline watcher/src/workers/supervisor.ts:19]: is this intended? Or should we use `logger`?

File: watcher/src/workers/supervisor.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  worker.on('exit', (code) => {
+    logger.warn(`Worker ${workerName} exited with code ${code}`);
+    if (code !== 0) {
+      logger.error(`Restarting worker ${workerName}...`);
+      spawnWorker(data);
+    }
+  });
+
+  workers[workerName] = { worker, data, lastHB: Date.now() };
+  console.log('Finished spawning worker:', workerName);
```
Comment (bingyuyap) [inline watcher/src/workers/supervisor.ts:38]: same as line 19!

File: watcher/src/workers/supervisor.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  data: WorkerData;
+  lastHB: number;
+}
+
+const workers: { [key: string]: WorkerInfo } = {};
+const logger = getLogger('supervisor');
+const network: Environment = getEnvironment();
+
+function spawnWorker(data: WorkerData) {
+  const workerName = `${data.chain}Worker`;
```
Comment (bingyuyap) [inline watcher/src/workers/supervisor.ts:18]: maybe let's log the network as well?

**Review by bingyuyap — APPROVED** [review]

File: watcher/src/workers/supervisor.ts (hunk 1 above)
Comment (panoel) [reply watcher/src/workers/supervisor.ts:19]: Fixed.

File: watcher/src/workers/supervisor.ts (hunk 3 above)
Comment (panoel) [reply watcher/src/workers/supervisor.ts:18]: Added

File: watcher/src/workers/supervisor.ts (hunk 2 above)
Comment (panoel) [reply watcher/src/workers/supervisor.ts:38]: Fixed

**Review by bingyuyap — APPROVED** [review]

**Pull request merged**